matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/DefaultCryptoService.kt

```diff
@@ -1371,6 +1376,54 @@ internal class DefaultCryptoService @Inject constructor(
         }
     }
 
+    override fun sendSharedHistoryKeysToLastChunk(roomId: String, userId: String, sessionInfoSet: Set<SessionInfoPair>?) {
+        cryptoCoroutineScope.launch(coroutineDispatchers.crypto) {
+            runCatching {
+                deviceListManager.downloadKeys(listOf(userId), false)
+            }.mapCatching {
+                val userDevices = cryptoStore.getUserDevices(userId)
+                userDevices?.forEach {
+                    // Lets share the provided inbound sessions for every user device
+                    val deviceId = it.key
+                    sessionInfoSet?.mapNotNull { sessionInfoPair ->
+                        // Get inbound session from sessionId and sessionKey
+                        cryptoStore.getInboundGroupSession(sessionInfoPair.first, sessionInfoPair.second)
+                    }?.filter { inboundGroupSession ->
+                        // Filter only sessions with sharedHistory enabled
+                        inboundGroupSession.sharedHistory
+                    }?.forEach { inboundGroupSession ->
+                        // Share the session to userId with deviceId
+                        val exportedKeys = inboundGroupSession.exportKeys()
+                        val algorithm = exportedKeys?.algorithm
+                        val decryptor = roomDecryptorProvider.getRoomDecryptor(roomId, algorithm)
+                        decryptor?.shareKeysWithDevice(exportedKeys, deviceId, userId)
+                        Timber.i("## CRYPTO | Sharing inbound session")
+                    }
+                }
+            }
+        }
+    }
+
+    override fun sendSharedHistoryKeys(roomId: String, userId: String) {
```

Comment:
the implementation of `sendSharedHistoryKeys` and `sendSharedHistoryKeysToLastChunk` looks very similar, can we share some of the logic?